### ISSUE TITLE
'kubectl version' returns invalid server semantic versioning

### ISSUE THREAD

**Issue by atoy3731** [rancher/rke2#513]
When running `kubectl version` against an RKE2 server, semantic versioning doesn't return as expected:

```
adam.toy@Adams-MacBook-Pro:~/P1/il5-argo-creds$ kubectl version
Client Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.6-beta.0", GitCommit:"e7f962ba86f4ce7033828210ca3556393c377bcc", GitTreeState:"clean", BuildDate:"2020-01-15T08:26:26Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"", Minor:"", GitVersion:"v1.18.10+rke2r1", GitCommit:"62876fc6d93e891aa7fbe19771e6a6c03773b0f7", GitTreeState:"clean", BuildDate:"2020-10-16T19:58:58Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
```

For my use-case, this specifically is affecting ArgoCD's ability to apply Helm charts.

**Comment by nikkelma** [comment]
The details about Argo appear more important than initially thought - it appears `gitMajor` and `gitMinor` variables in versioning metadata are deprecated, indicating that apps should only be parsing `gitVersion` for information [1].

Should the deprecated compile-time flags be included to maximize backwards compatibility, or should this be pushed forward as the responsibility of applications/operators to use the new functionality?

[1] `kubernetes/component-base` branch `release-1.18` as of this comment: https://github.com/kubernetes/component-base/blob/65f5cb33b5afd6f30c48104cb5f721a97b0b0076/version/base.go#L37-L43

**Comment by cjellick** [comment]
@briandowns is this merged? if so can you attach to the pr and put in the appropriate milestone (ask @davidnuzik for that).

**Comment by briandowns** [comment]
#500 has been merged.

**Comment by rancher-max** [comment]
#### Tested using `v1.18.10-alpha1+rke2r1` - Failing for Client Version

```
$ kubectl version
Client Version: version.Info{Major:"", Minor:"", GitVersion:"v1.18.10-alpha1+rke2r1", GitCommit:"62876fc6d93e891aa7fbe19771e6a6c03773b0f7", GitTreeState:"clean", BuildDate:"2020-11-05T18:08:57Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.10-alpha1+rke2r1", GitCommit:"62876fc6d93e891aa7fbe19771e6a6c03773b0f7", GitTreeState:"clean", BuildDate:"2020-11-05T18:05:13Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
```

Notice the Major and Minor versions show for the server version, but are not showing for client version. I validated that in k3s both the major and minor versions are showing for both client and server.

Reopening since this is failing for client version when using the kubectl installed with rke2.

**Comment by dweomer** [comment]
Not sure what is going on here:

`make dev-shell`:
```
kubectl version
Client Version: version.Info{Major:"", Minor:"", GitVersion:"v1.18.10-alpha1+rke2", GitCommit:"62876fc6d93e891aa7fbe19771e6a6c03773b0f7", GitTreeState:"clean", BuildDate:"2020-11-16T20:10:41Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.10-alpha1+rke2", GitCommit:"62876fc6d93e891aa7fbe19771e6a6c03773b0f7", GitTreeState:"clean", BuildDate:"2020-11-16T20:05:33Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
```

`docker run --rm -it rancher/hardened-kubernetes:v1.18.10-alpha1-rke2 kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.10-alpha1+rke2", GitCommit:"62876fc6d93e891aa7fbe19771e6a6c03773b0f7", GitTreeState:"clean", BuildDate:"2020-11-16T20:05:33Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
The connection to the server localhost:8080 was refused - did you specify the right host or port?
```

**Comment by davidnuzik** [comment]
@briandowns this can be next up after K3s CIS work.

**Comment by rancher-max** [comment]
Validated fixed on master build from commit `5236c5b9`. Steps were:
```
$ make

$ mkdir -p /var/lib/rancher/rke2/agent/images && \
cp bin/rke2 /usr/local/bin/rke2 && \
cp ./build/images/rke2-runtime.tar /var/lib/rancher/rke2/agent/images/ && \
cp ./build/images/rke2-kubernetes.tar /var/lib/rancher/rke2/agent/images/

$ sudo rke2 server --write-kubeconfig-mode 644 2>&1 &

$ export KUBECONFIG=/etc/rancher/rke2/rke2.yaml PATH=$PATH:/var/lib/rancher/rke2/bin
$ kubectl version
```

Result:
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.12-dev+5236c5b9", GitCommit:"7cd5e9086de8ae25d6a1514d0c87bac67ca4a481", GitTreeState:"clean", BuildDate:"2020-12-11T16:52:47Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.12-dev+5236c5b9", GitCommit:"7cd5e9086de8ae25d6a1514d0c87bac67ca4a481", GitTreeState:"clean", BuildDate:"2020-12-11T16:52:47Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
```

I'd like to give this a quick check on an rc before closing due to the differences Jacob saw when running `make dev-shell` vs running the hardened-kubernetes image.

**Comment by rancher-max** [comment]
Validated on v1.18.13-rc1+rke2r1 as well, this is working:
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.13-rc1+rke2r1", GitCommit:"4c00c3c459261e8ff3381c1070ddf798f0131956", GitTreeState:"clean", BuildDate:"2020-12-11T21:09:43Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.13-rc1+rke2r1", GitCommit:"4c00c3c459261e8ff3381c1070ddf798f0131956", GitTreeState:"clean", BuildDate:"2020-12-11T21:09:43Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
```
